### PR TITLE
Rewrite and refactor all of pycoast to meet minimum modern standards

### PR DESCRIPTION
**Pull request by djhoese**
This PR makes every change necessary to make pycoast pass basic formatting, linting, and complexity checks. To be sure of that I have enabled pre-commit.ci and added a pre-commit config that can be used by developers to ensure the quality of the code moving forward. I am almost done with the flake8 complexity issues and then there are a few bandit security issues I need to look into.

 - [ ] Closes #xxxx 
 - [ ] Tests added 
 - [ ] Tests passed 
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
This could still definitely use a lot more work, but I've reduce the complexity *a lot* for the entire `cw_base.py` module. Some of it still isn't pretty, but I think it requires considering changing the algorithm. The hard part for handling the grid lines was that the minor lines depend on what the major lines are and a couple other odd combinations of things. This made it hard to separate the code further.

@lobsiger CodeFactor is *finally* happy :wink:

File: docs/source/test.rst [hunk 1]
```diff
@@ -6,5 +6,4 @@ The tests can be run using nosetest:
 .. code-block:: bash
 
     $ cd <pycoast_dir>
-    $ nosetests tests/  
-
+    $ nosetests tests/
```
Comment (mraspaud) [inline docs/source/test.rst:9]: Are we still using nose?

File: pycoast/cw_base.py [hunk 2]
```diff
@@ -144,419 +152,89 @@ class ContourWriterBase(object):
...
     def _find_bounding_box(self, xys):
         lons = [x for (x, y) in xys]
         lats = [y for (x, y) in xys]
         return [min(lons), min(lats), max(lons), max(lats)]
 
-    def _add_shapefile_shapes(self, image, area_def, filename,
-                              feature_type=None, **kwargs):
+    def _add_shapefile_shapes(
+        self, image, area_def, filename, feature_type=None, **kwargs
+    ):
```
Comment (mraspaud) [inline pycoast/cw_base.py:218]: can this be put on one line?

File: pycoast/cw_base.py [hunk 3]
```diff
@@ -144,419 +152,89 @@ class ContourWriterBase(object):
...
         """Draw all shapes (polygon/poly-lines) from a shape file onto a PIL Image."""
         sf = shapefile.Reader(filename)
-        return self.add_shapes(image, area_def, sf.shapes(), feature_type=feature_type, **kwargs)
+        return self.add_shapes(
+            image, area_def, sf.shapes(), feature_type=feature_type, **kwargs
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:223]: Can this be put one line?

File: pycoast/cw_base.py [hunk 4]
```diff
@@ -144,419 +152,89 @@ class ContourWriterBase(object):
...
 
-    def _add_shapefile_shape(self, image, area_def, filename, shape_id,
-                             feature_type=None, **kwargs):
+    def _add_shapefile_shape(
+        self, image, area_def, filename, shape_id, feature_type=None, **kwargs
+    ):
```
Comment (mraspaud) [inline pycoast/cw_base.py:227]: Can this be put on one line?

File: pycoast/cw_base.py [hunk 5]
```diff
@@ -144,419 +152,89 @@ class ContourWriterBase(object):
...
 
         Accesses single shape using shape_id from a custom shape file.
 
         """
         sf = shapefile.Reader(filename)
         shape = sf.shape(shape_id)
-        return self.add_shapes(image, area_def, [shape], feature_type=feature_type, **kwargs)
+        return self.add_shapes(
+            image, area_def, [shape], feature_type=feature_type, **kwargs
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:237]: Can this be put back on one line?

File: pycoast/cw_base.py [hunk 6]
```diff
@@ -638,113 +324,163 @@ def add_shapes(self, image, area_def, shapes, feature_type=None, x_offset=0, y_o
...
-                                 (db_name, resolution, i))
-            else:
-                shapefilename = \
-                    os.path.join(db_root_path, '%s_shp' % db_name,
-                                 resolution, format_string %
-                                 (db_name, tag, resolution, i))
+        levels = range(1, level + 1) if not isinstance(level, list) else level
+        format_string, format_params = self._get_db_shapefile_format_and_params(
+            db_name, resolution, tag, zero_pad
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:456]: This fits in one line

File: pycoast/cw_base.py [hunk 7]
```diff
@@ -638,113 +324,163 @@ def add_shapes(self, image, area_def, shapes, feature_type=None, x_offset=0, y_o
...
+        shapefile_root_dir = os.path.join(db_root_path, f"{db_name}_shp", resolution)
+        for i in levels:
+            level_format_params = format_params + (i,)
+            shapefilename = os.path.join(
+                shapefile_root_dir, format_string % level_format_params
+            )
```
Comment (mraspaud) [inline pycoast/cw_base.py:462]: This fits on one line

File: pycoast/cw_base.py [hunk 8]
```diff
@@ -778,7 +522,9 @@ def _config_to_dict(self, config_file):
                         overlays[section][option] = val
         return overlays
 
-    def add_overlay_from_dict(self, overlays, area_def, cache_epoch=None, background=None):
+    def add_overlay_from_dict(
+        self, overlays, area_def, cache_epoch=None, background=None
+    ):
```
Comment (mraspaud) [inline pycoast/cw_base.py:527]: This fits on one line

File: pycoast/cw_base.py [hunk 9]
```diff
@@ -823,233 +569,10 @@ def add_overlay_from_dict(self, overlays, area_def, cache_epoch=None, background
...
-                continue
-            params_to_hash[overlay_name] = overlay_dict.copy()
-        # font objects are not hashable
-        for font_cat in ('cities', 'points', 'grid'):
-            if font_cat in params_to_hash:
-                params_to_hash[font_cat].pop('font', None)
-        return params_to_hash
+        overlay_helper = _OverlaysFromDict(
+            self, overlays, area_def, cache_epoch, background
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:574]: This fits on one line.

File: pycoast/cw_base.py [hunk 10]
```diff
@@ -1062,8 +585,9 @@ def add_overlay_from_config(self, config_file, area_def, background=None):
 
         """
         overlays = self._config_to_dict(config_file)
-        return self.add_overlay_from_dict(overlays, area_def,
-                                          os.path.getmtime(config_file), background)
+        return self.add_overlay_from_dict(
+            overlays, area_def, os.path.getmtime(config_file), background
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:590]: This should fit on one line

File: pycoast/cw_base.py [hunk 11]
```diff
@@ -85,15 +87,19 @@ def __init__(self, coord_ref: str, area_def: AreaDefinition):
             "image": self._image_to_pixels,
         }
         if coord_ref not in convert_methods:
-            pretty_coord_refs = [f"'{cr_name}'" for cr_name in sorted(convert_methods.keys())]
+            pretty_coord_refs = [
+                f"'{cr_name}'" for cr_name in sorted(convert_methods.keys())
+            ]
```
Comment (mraspaud) [inline pycoast/cw_base.py:92]: This fits on one line with 120 characters

File: pycoast/cw_base.py [hunk 12]
```diff
@@ -144,419 +152,89 @@ class ContourWriterBase(object):
...
-                    continue
-
-                # make PIL draw the lines...
-                for index_array in index_arrays:
-                    self._draw_line(draw,
-                                    index_array.flatten().tolist(),
-                                    **kwargs)
+        grid_drawer = _GridDrawer(
+            self, draw, area_def, Dlon, Dlat, dlon, dlat, font, write_text, kwargs
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:207]: This should fit on one line

File: pycoast/cw_base.py [hunk 13]
```diff
@@ -623,10 +310,9 @@ def add_shapes(self, image, area_def, shapes, feature_type=None, x_offset=0, y_o
         prj = Proj(proj_def)
 
         # Calculate min and max lons and lats of interest
-        lon_min, lon_max, lat_min, lat_max = _get_lon_lat_bounding_box(area_extent,
-                                                                       x_size,
-                                                                       y_size,
-                                                                       prj)
+        lon_min, lon_max, lat_min, lat_max = _get_lon_lat_bounding_box(
+            area_extent, x_size, y_size, prj
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:315]: Fits on one line

File: pycoast/cw_base.py [hunk 14]
```diff
@@ -638,113 +324,163 @@ def add_shapes(self, image, area_def, shapes, feature_type=None, x_offset=0, y_o
...
+            if len(index_arrays) == 0:
+                return
+
+            # Make PIL draw the polygon or line
+            for index_array in index_arrays:
+                if ftype == "polygon" and not is_reduced:
+                    # Draw polygon if dataset has not been reduced
+                    self._draw_polygon(
+                        draw, index_array.flatten().tolist(), **new_kwargs
+                    )
```
Comment (mraspaud) [inline pycoast/cw_base.py:398]: one line

File: pycoast/cw_base.py [hunk 15]
```diff
@@ -638,113 +324,163 @@ def add_shapes(self, image, area_def, shapes, feature_type=None, x_offset=0, y_o
...
+        x_offset=0,
+        y_offset=0,
+        db_root_path=None,
+        **kwargs,
+    ):
         """Add a contour feature to image."""
         shape_generator = self._iterate_db(
-            db_name, tag, resolution, level, zero_pad,
-            db_root_path=db_root_path
+            db_name, tag, resolution, level, zero_pad, db_root_path=db_root_path
```
Comment (mraspaud) [inline pycoast/cw_base.py:434]: one line, right?

File: pycoast/cw_base.py [hunk 16]
```diff
@@ -1162,104 +718,95 @@ def add_cities(self, image, area_def, cities_list, font_file, font_size=12,
         try:
             from pyresample.geometry import AreaDefinition
         except ImportError:
-            raise ImportError("Missing required 'pyresample' module, please install it.")
+            raise ImportError(
+                "Missing required 'pyresample' module, please install it."
+            )
```
Comment (mraspaud) [inline pycoast/cw_base.py:723]: 1 line

File: pycoast/cw_base.py [hunk 17]
```diff
@@ -1162,104 +718,95 @@ def add_cities(self, image, area_def, cities_list, font_file, font_size=12,
...
 
         if not isinstance(area_def, AreaDefinition):
-            raise ValueError("Expected 'area_def' is an instance of AreaDefinition object")
+            raise ValueError(
+                "Expected 'area_def' is an instance of AreaDefinition object"
+            )
```
Comment (mraspaud) [inline pycoast/cw_base.py:728]: I think you get my point now, one line

File: pycoast/cw_base.py [hunk 18]
```diff
@@ -1162,104 +718,95 @@ def add_cities(self, image, area_def, cities_list, font_file, font_size=12,
...
-
-        for city_row in cities_file:
-            city_info = city_row.split('\t')
-            if not city_info or not (city_info[1] in cities_list or city_info[2] in cities_list):
-                continue
-            city_name, lon, lat = city_info[1], float(city_info[5]), float(city_info[4])
-
+        cities_filename = os.path.join(
+            db_root_path, os.path.join("CITIES", "cities.txt")
+        )
```
Comment (mraspaud) [inline pycoast/cw_base.py:736]: guess what?

File: pycoast/cw_base.py [hunk 19]
```diff
@@ -1324,138 +871,203 @@ def add_points(self, image, area_def, points_list, font_file, font_size=12,
...
-
-                    box_outline = new_kwargs.pop('box_outline', 'white')
-                    box_opacity = new_kwargs.pop('box_opacity', 0)
-
-                    # add text_box
-                    self._draw_text_box(draw, text_position, desc, font, outline,
-                                        box_outline, box_opacity, **new_kwargs)
+                logger.info(
+                    f"Point ({x}, {y}) is out of the image area, it will not be added to the image."
+                )
```
Comment (mraspaud) [inline pycoast/cw_base.py:876]: here too

File: pycoast/cw_base.py [hunk 20]
```diff
@@ -1324,138 +871,203 @@ def add_points(self, image, area_def, points_list, font_file, font_size=12,
...
+                    ptsize,
+                    outline,
+                    outline_opacity,
+                    fill,
+                    fill_opacity,
+                )
+            elif desc is None:
+                logger.error(
+                    "'ptsize' is 0 and 'desc' is None, nothing will be added to the image."
+                )
```
Comment (mraspaud) [inline pycoast/cw_base.py:906]: too short to be split

File: pycoast/cw_base.py [hunk 21]
```diff
@@ -1324,138 +871,203 @@ def add_points(self, image, area_def, points_list, font_file, font_size=12,
...
-            continue
-        if prev is not None:
-            delta = lon - prev
-            if abs(delta) > 180:
-                delta = (abs(delta) - 360) * np.sign(delta)
-            angle_sum += delta
-        prev = lon
+    bbox_lons, bbox_lats = _get_bounding_box_lonlat_sides(
+        area_extent, x_size, y_size, prj
+    )
```
Comment (mraspaud) [inline pycoast/cw_base.py:1039]: seriously?

File: pycoast/cw_base.py [hunk 22]
```diff
@@ -1538,3 +1182,686 @@ def _get_pixel_index(shape, area_extent, x_size, y_size, prj,
...
+        self._cities_file = open(cities_filename, mode="r", encoding="utf-8")
+
+    def iter_cities_names_lon_lat(
+        self, cities_list: list[str]
+    ) -> Generator[tuple[str, float, float], None, None]:
+        for city_row in self._cities_file:
+            city_info = city_row.split("\t")
+            if not city_info or not (
+                city_info[1] in cities_list or city_info[2] in cities_list
+            ):
```
Comment (mraspaud) [inline pycoast/cw_base.py:1200]: This is badly split.

File: pycoast/cw_base.py [hunk 23]
```diff
@@ -1538,3 +1182,686 @@ def _get_pixel_index(shape, area_extent, x_size, y_size, prj,
...
+        self._is_cached = False
+        self._cache_filename = None
+        self._background = background
+        self._area_def = area_def
+
+        foreground = None
+        if "cache" in overlays:
+            cache_filename, foreground = self._get_cached_filename_and_foreground(
+                cache_epoch
+            )
```
Comment (mraspaud) [inline pycoast/cw_base.py:1221]: one line

**Review by mraspaud — CHANGES_REQUESTED** [review]
Ok, I got tired of writing the same comment over and over, I think black should be fixed to allow 120 characters and not split stuff randomly.

**Comment by djhoese** [comment]
Ok, your comments have been addressed. I mentioned pytest in the test document. I updated black from using its default 88 character line length to 120.

**Comment by mraspaud** [comment]
> Ok, your comments have been addressed. I mentioned pytest in the test document. I updated black from using its default 88 character line length to 120.

Great, I can continue reviewing then :) Thanks!!!

**Review by mraspaud — APPROVED** [review]
LGTM. Mixing the refactoring with "blackification" didn't help either, maybe the next time we do this, we should do two separate PRs? But from what I can see, the tests didn't change in functionality, so I think we're good.

**Comment by djhoese** [comment]
Yes, agreed. This started as a "add pre-commit" PR but to include all of the hooks I wanted it required rewriting most of the package. And yes, no behavior was intentionally changed. Merging now and then I could start working toward a config interface and automatically download the coastline files.

**Comment by lobsiger** [comment]
@djhoese @mraspaud  now that this refactoring PR is merged, do we have a chance to see a Pycoast V. 1.6.0 anytime soon?

**Comment by djhoese** [comment]
@lobsiger I'm currently finishing my recovery from COVID and will need to submit a final report for one of my projects next week *then* catch up on all my other work. After all that, yes I'd like to look into a release. However, I noticed in the newest pre-commit PR #64 that one of the tests is failing with the new versions of dependencies. This test wasn't failing when this PR last ran it's CI. Debugging this test failure is the first major hurdle to getting a new release out. If you have any free time to investigate it would be much appreciated. Just narrowing the problem down to what versions of what dependencies causes the failure would be **huge**.

**Pull request merged**